go/libraries/doltcore/sqle/autoincr/auto_increment_tracker.go

```diff
@@ -0,0 +1,120 @@
+package autoincr
+
+import (
+	"fmt"
+	"sync"
+)
+
+type AutoIncrementTracker interface {
```

Comment:
This is the wrong interface. It's combining two different patterns.

Either this thing should be dumb, in which case its only responsibility is to tell people which IDs have been passed out already, and let them tell it "hey I passed out this ID"

Or it should be smart, and handle passing out the Next ID automatically.

What you have here combines these two modes of operation, and it's kind of a mess.

I would recommend going with option 2, and making this thing smart. Make it responsible for handing out all auto increment keys for each table.

That leaves you with just two interface methods here: InitTable(tableName, initialVal) and Next(tableName). Next should return a specific error type if that table hasn't been initialized yet in this session, and then the client can call InitTable on it.

The above pattern means you have to be tolerant of multiple init calls. You need to make InitTable forgiving if it gets called by more than one thread (probably silently succeeding if gets called more than once). The other option is to add Lock and Unlock methods to the interface so that callers can do the above two calls in a single mutex.

Reply:
Also you don't need to worry about skipping or burning auto increment IDs that you handed out to transactions that eventually fail. MySQL does this too, and application developers already have to deal with it.